File: bonsai_scxml_web/src/main/java/de/unibi/citec/clf/bonsai/engine/communication/web/WebController.java
package de.unibi.citec.clf.bonsai.engine.communication.web;


import de.unibi.citec.clf.bonsai.engine.communication.SCXMLRemote;
import de.unibi.citec.clf.bonsai.engine.control.StateMachineController;

import java.util.List;
import java.util.Map;

/**
 * @author lruegeme
 */
public class WebController implements SCXMLRemote {

    private static final org.apache.log4j.Logger logger = org.apache.log4j.Logger.getLogger(WebController.class);

    @Override
    public boolean fireEvent(String event) {
        return false;
    }

    @Override
    public List<String> getCurrentStates() {
        return List.of();
    }

    @Override
    public List<String> getStateIds() {
        return List.of();
    }

    @Override
    public List<String> getTransitions() {
        return List.of();
    }

    @Override
    public String load(String pathToConfig, String pathToTask, Map<String, String> includeMapping, boolean forceConfigure) {
        return "";
    }

    @Override
    public boolean pause() {
        return false;
    }

    @Override
    public boolean resume() {
        return false;
    }

    @Override
    public boolean setParams(Map<String, String> map) {
        return false;
    }

    @Override
    public boolean start() {
        return false;
    }

    @Override
    public boolean start(String state) {
        return false;
    }

    @Override
    public boolean stop() {
        return false;
    }

    @Override
    public boolean stopAutomaticEvents(boolean b) {
        return false;
    }

    @Override
    public void exit() {

    }
}

File: bonsai_scxml_web/src/main/java/de/unibi/citec/clf/bonsai/engine/communication/web/WebController.kt
package de.unibi.citec.clf.bonsai.engine.communication.web

import de.unibi.citec.clf.bonsai.engine.communication.SCXMLRemote
import de.unibi.citec.clf.bonsai.engine.communication.web.model.LoadData
import de.unibi.citec.clf.bonsai.engine.communication.web.model.StateIds
import de.unibi.citec.clf.bonsai.engine.communication.web.model.Transitions
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.request.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*
import kotlinx.coroutines.runBlocking
import org.apache.log4j.Logger


/**
 * @author lruegeme
 */
open class WebController(val host: String, val port: Int) : SCXMLRemote {
    val bonsai : String = "http://$host:$port/bonsai"
    val client : HttpClient = HttpClient() {
        install(ContentNegotiation) {
            json()
        }
    }

    override fun fireEvent(event: String): Boolean {
        runBlocking {
            client.post("$bonsai/fire_event") {
                contentType(ContentType.Application.Json)
                setBody(event)
            }
        }
        return true
    }

    override fun getCurrentStates(): MutableList<String> {
        val response : StateIds = runBlocking {
            client.get("$bonsai/states").body()
        }
        return response.ids as MutableList<String>
    }

    override fun getStateIds(): MutableList<String> {
        val response : StateIds = runBlocking {
            client.get("$bonsai/all_states").body()
        }
        return response.ids as MutableList<String>
    }

    override fun getTransitions(): MutableList<String> {
        val response : Transitions =  runBlocking {
            client.get("$bonsai/transitions").body()
        }
        return response.transitions as MutableList<String>
    }

    override fun load(
        pathToConfig: String,
        pathToTask: String,
        includeMapping: MutableMap<String, String>,
        forceConfigure: Boolean
    ): String {
        val ld = LoadData(pathToConfig, pathToTask, includeMapping, forceConfigure)
        val response =  runBlocking {
            client.post("$bonsai/load") {
                contentType(ContentType.Application.Json)
                setBody(ld)
            }
        }
        return response.toString()
    }

    override fun pause(): Boolean {
        return false
    }

    override fun resume(): Boolean {
        return false
    }

    override fun setParams(map: MutableMap<String, String>?): Boolean {
        return false
    }

    override fun start(): Boolean {
        runBlocking {
            client.post("$bonsai/start")
        }
        return true
    }

    override fun start(state: String): Boolean {
        runBlocking {
            client.post("$bonsai/start") {
                setBody(state)
            }
        }
        return true
    }

    override fun stop(): Boolean {
        runBlocking {
            client.post("$bonsai/stop") {
            }
        }
        return true
    }

    override fun stopAutomaticEvents(b: Boolean): Boolean {
        runBlocking {
            client.post("$bonsai/stop_events") {
                contentType(ContentType.Application.Json)
                setBody(b)
            }
        }
        return true
    }

    override fun exit() {
    }

    companion object {
        private val logger: Logger? = Logger.getLogger(WebController::class.java)
    }
}
